### ISSUE TITLE
Trigger FirebaseInAppMessagingClickListener. messageClicked when action is empty

### ISSUE THREAD

**Issue by KonradSzewczuk** [firebase/firebase-android-sdk#1603]
## What feature would you like to see?
Invoke FirebaseInAppMessagingClickListener.messageClicked for use case when you don't want to trigger browser intent and you are interested only to handle other data (such as key-value map defined in the campaign)

As a user, I don't want to trigger any URL specific browser intent when i click my In App Message popup. I want to trigger below code without any side effects:

```val inAppMessaging = Firebase.inAppMessaging
inAppMessaging.addClickListener { inAppMessage, action ->
      // invoke only code inside
          val data = inAppMessage.data // I am only interested in that data, don't want to trigger any browser intent
      }
```

Unfortunately, this listener is not being invoked when the defined action is empty. You must define the action with proper URL to be able to trigger this listener when popup is clicked.

Can we trigger such listener always when the popup is being clicked, even when the defined action for campaign is empty?

**Comment by JasonAHeron** [comment]
Hey @KonradSzewczuk, this is an interesting feature request. Right now when a user does not provide a valid click action we treat clicks as dismiss and trigger your dismiss callback but I can see how you might want to differentiate between a user actually clicking to dismiss vs clicking on the FIAM.

So what you're imagining is if an action is not set then when a FIAM is clicked we call the developer registered callback for click and dismiss? rather than dismiss like we do right now?

**Comment by ashwinraghav** [comment]
Thanks for the feature request.
If you are looking for programmatic invocation of remote functions defined in your app, consider using [FirebaseMessaging](https://firebase.google.com/docs/cloud-messaging).
You could [send](https://firebase.google.com/docs/cloud-messaging/http-server-ref) a message to your app from your app's backend or from the Notifications tab in the [Firebase console](https://console.firebase.google.com/).
Upon [receiving](https://firebase.google.com/docs/cloud-messaging/android/receive) the message, your app can parse it and invoke any functions that you would like.

**Comment by KonradSzewczuk** [comment]
@JasonAHeron I can confirm that your description is something that I am looking for.

In addition, my use case is to trigger a InAppMessaging popup and after click on it change the Fragment in my activity without any current activity refresh. To achieve that I think I must not trigger any deeplink related browser intent. The easiest way would be to handle such scenario in the FirebaseInAppMessagingClickListener. messageClicked  as I proposed (without any side effects)
I don't see any possible reason to use FirebaseMessaging in that case as @ashwinraghav proposed.

@ashwinraghav @JasonAHeron Can we reopen this feature request?

**Comment by ashwinraghav** [comment]
Hey @KonradSzewczuk,
Closed the ticket unintentionally. Sry bout that. Also, I may have misunderstood what you are reporting.
Thanks for clarifying.
@JasonAHeron seems like a change that is non breaking IMO. Seems safe to have this roll out just as a minor version ? Also worth checking what the behavior on iOS is.

**Comment by PiotrPrus** [comment]
The feature that we are talking about is a dismissal event. I have checked the documentation for iOS and in the CardActionFiamDisplay there is a messageDismissed listener: https://firebase.google.com/docs/in-app-messaging/modify-message-behavior?platform=ios#implement_a_displaydelegate_to_handle_card_interactions
As far I as understand, the same functionality on android would resolve the issue. Why these two even do not share the same functionalities? :(

**Comment by 8kt8** [comment]
I am straggling with the same problem. 
I try to put my custom url like service://book and open proper book screen in my app after button click action. 
Now it is impossible, app crash: 
No Activity found to handle Intent { act=android.intent.action.VIEW dat=service://book (has extras) }  
I my honest opinion is basic functionality to do something in the application after click button on In-AppMessage, without triggering any intent to the browser.   
So +1 from my site for this feature request.
Any suggestions for a workaround? I won't dismiss the in-app message and change fragment.

**Comment by JasonAHeron** [comment]
Hey folks, thanks for bringing this to my attention. I'm going to prioritize this this week and hopefully have a fix in the next release.

**Comment by JasonAHeron** [comment]
Okay so after reading through this issue again it appears there are actually three different requests here all wrapped up in one issue.

1. @KonradSzewczuk wants to know if it would be possible to trigger the clicked listener rather than just a dismiss if the action url is not provided

2. @PiotrPrus is asking for a dismiss developer callback

3. @8kt8 is asking about handling cases where the developer has specified a URL action but the user has no browser installed

Do I have all these requests correct?

**Comment by 8kt8** [comment]
Hi @JasonAHeron thanks for the quick response.
Firebase InAppMessaging should not send Intent with custom scheme of data.
Now I am trying to set button action like myAppId://sceenA, myAppId://sceenB, myAppId://sceenC.
After trigger action by the user, I want only to change a screen inside the app.
Now my app crashing: No Activity found to handle Intent { act=android.intent.action.VIEW dat=myAppId://sceenA (has extras) }
FirebaseInAppMessagingClickListener should provide the option to disable the sending of intent or give a possibility to override this default behavior.
In my case sending Intent is not needed. 
The same problem is when I want to open the Uri inside my app in WebView instead of the external browser.

**Comment by JasonAHeron** [comment]
@8kt8 so my understanding is this could be supported if we call the click handler when a button is clicked with an empty action url. Then you could populate your custom data using the custom data bundles field in the campaign.

This would solve what both you and @KonradSzewczuk are asking for

**Comment by 8kt8** [comment]
@JasonAHeron Your proposal almost solves my problems. You mention about empty action url,  what in case with opening url action in WebView in application instead of opening url in external browser?

**Comment by KonradSzewczuk** [comment]
> Okay so after reading through this issue again it appears there are actually three different requests here all wrapped up in one issue.
> 
> 1. @KonradSzewczuk wants to know if it would be possible to trigger the clicked listener rather than just a dismiss if the action url is not provided
> 2. @PiotrPrus is asking for a dismiss developer callback
> 3. @8kt8 is asking about handling cases where the developer has specified a URL action but does not want it handled by the browser.
> 
> Do I have all these requests correct?

@JasonAHeron I think clickListener should always be invoked when we click on popup. The problem is it should not be somehow connected with specifying obligatory browser intent action in the console and handling it. This will cover the usecase when I am only interested in the key-value data defined in the console and for example I want to change Fragment in my app after the popup is being clicked. So I will not be forced to trigger the browser intent.
Having a dismiss click listener is something different, I found it also useful but it is not connected with my story.
So it would be great if both those features will be added

**Comment by JasonAHeron** [comment]
@KonradSzewczuk Yeah makes sense, I'm going to bring this request to the team and make sure we can be consistent with it across iOS and Android and our docs but it does make sense and I'm going to push to make it happen.

@8kt8 my understanding is if you want to handle a click in your app via your own mechanism then you can do that (assuming the click handler is always called) by just adding whatever data you want into the data bundle and then handling that however you want in your click handler. This would work for launching your own web view instead of the browser

To summarize I think we have two good feature requests here

1. Always call the click listener even when there is no action url
2. Add a dismiss listener to be consistent with iOS and better support developer interactivity

**Comment by 8kt8** [comment]
@JasonAHeron Yes, I can put my custom data to Bundle, but there is a difference between iOS and Android.
On iOS is possible to add custom button action like appId://screenA, handle custom on click listener. iOS doesn't trigger an external browser.
On Android in the same scenario system send an intent with url, which causes crashes for custom url scheme if the proper intent filter is not added to app manifest(Maybe you should catch NoActivityException and not crash the app). 
The best solution would be to have the same behavior on Android.

**Comment by 8kt8** [comment]
Please consider a case with more than one button on InAppMessaging Dialog.
Now the only way to recognize which button was clicked is to compare action urls.
In my opinion, when the developer define custom on click handler, default behavior with sending url intent should be overridden.

**Comment by JasonAHeron** [comment]
Okay so this is a request for something like "Prevent Default" on click listeners. Makes sense. I'll bundle it into the proposal for click behavior changes

**Comment by JasonAHeron** [comment]
The above change should improve a few things
1) Add a dismiss callback to be consistent with iOS
2) Graceful handling for launching a URI intent with the user has no browser installed
3) Now if a developer registers a click callback in the SDK then they will always get a chance to react to clicks even if that click does not have a URL.

**Comment by shaozl** [comment]
Hi @JasonAHeron @8kt8 , I have the same issue with @8kt8 , in my app , we will receive dynamic URI through In-app messaging's FirebaseInAppMessagingClickListener. messageClicked() , we decided to analysis the URI scheme://hosts?querymaps and get the queryMaps then do dome jobs. Unfortunately, our app crash with 'No Activity found to handle Intent { act=android.intent.action.VIEW ', after reading In-app messaging i found the Intent you guys 'hardcode' there with our customized URI.
So from my site 
       1. We passing data through the URI ,  maybe with dynamic scheme, could you guys help to remove that Hardcode Intent , just leave the ClickListener for the developers , many thx
       2. In our app we send in-app messaging to let our customer know ,if there are some big events, when the user clicks on the Message , they will navigate to a specific page.   Maybe you can provide more types than Uri, that will improve developers experience,thx

**Comment by tvietinghoff** [comment]
Hi @JasonAHeron, 
thank you for the improved handling regarding calling click listeners with empty actionUrls. However, the Firebase console does not let me specify an empty actionUrl for the Modal  message layout, so this work-around is not helping my case.

I think the more universal approach would be a "prevent default" option for all listeners. But if the console could be modified to allow the empty URL be handled as dismiss just like with the Card layout, it would be sufficient for these cases for now and would not require an API change.

![image](https://user-images.githubusercontent.com/3853604/100862917-178f3f00-3494-11eb-8bc9-826b3cd9e190.png)

**Comment by Kranthi94** [comment]
Still not working
